### PR TITLE
DO NOT MERGE RFC Interface for module splitting

### PR DESCRIPTION
**Pull request by gzm0**
If you come here as an sbt user, in this design, `ModuleOutputSpec` is what you'd have to provide in your build (`ModuleLinkerOutput` would probably be calculated automatically).

### CODE REVIEW

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Linker.scala [hunk 1]
```diff
@@ -23,8 +23,21 @@ import org.scalajs.logging.Logger
...
       moduleInitializers: Seq[ModuleInitializer],
       output: LinkerOutput, logger: Logger)(
+      implicit ec: ExecutionContext): Future[Unit] = {
+    val module = ModuleOutputSpec.Module("core")
+      .withInitializers(moduleInitializers: _*)
+
+    val spec = ModuleOutputSpec()
+      .addModule(ModuleOutputSpec.Selector.default, module)
+
+    val moduleOutput = ModuleLinkerOutput(module.id, output)
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Linker.scala]: Note: By implicitly defining that an external module is enough, if it does not have any dependencies, we can reduce the base case to a module case.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Linker.scala [hunk 2]
```diff
@@ -23,8 +23,21 @@ import org.scalajs.logging.Logger
...
+
+    link(irFiles, spec, Seq(moduleOutput), logger)
+  }
+
+  def link(irFiles: Seq[IRFile], moduleSpec: ModuleOutputSpec, outputs: Seq[ModuleLinkerOutput], logger: Logger)(
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Linker.scala]: Note: The reason `moduleSpec` here is separate from `outputs` is that I anticipate most linker usages to be in a build system. In such a system, `moduleSpec` is provided by the user, but `outputs` is provided by the build system (probably dependent on `moduleSpec`).

This is less nice, because it will cause runtime failures, if the outputs are not in sync with what is in the `moduleSpec`.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ *  @param isInternal Whether this is an internal (or an external) module.
+ *
+ *  @param output The [[LinkerOutput]] for this module.
+ *
+ *  @param moduleName Name of this module for import. A [[Linker]] will use
+ *      this name to include this module from one of its dependees. Must be set,
+ *      if this module is internal.
+ */
+final class ModuleLinkerOutput private (
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala]: In another design, I have merged these fields into `LinkerOuput`. I'm not 100% sure what is better.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleOutputSpec.scala [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.linker.interface
+
+/** Module output specification. */
+final class ModuleOutputSpec private (modules: Seq[(ModuleOutputSpec.Selector, ModuleOutputSpec.Module)]) {
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleOutputSpec.scala]: We need to finely balance what is accessible in this interface. The use case of a build tool would require access to the list of modules to prepare the relevant output files.

However, `deps` and `initializers` should probably not be accessible. (if we are to make `deps` accessible, we should make the verified and potentially inferred deps accessible).

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleOutputSpec.scala [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      val id: ModuleID,
+      initializers: Seq[ModuleInitializer],
+      deps: Seq[ModuleID]
+  ) {
+    private def this(id: ModuleID) = this(id, Nil, Nil)
+
+    def withInitializers(initializers: ModuleInitializer*): Module =
+      copy(initializers = initializers)
+
+    def withDeps(deps: ModuleID*): Module =
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleOutputSpec.scala]: The idea here is that in a future version we can add `withAutoDeps` if this is a feature that is needed.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleOutputSpec.scala [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.linker.interface
+
+/** Module output specification. */
+final class ModuleOutputSpec private (modules: Seq[(ModuleOutputSpec.Selector, ModuleOutputSpec.Module)]) {
+  private def this() = this(Nil)
+
+  def addModule(selector: ModuleOutputSpec.Selector, module: ModuleOutputSpec.Module): ModuleOutputSpec = {
```
Comment (gzm0) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleOutputSpec.scala]: The idea here is that in a future version we can add: `addAutoClassifyModule(selector)` if this feature is needed.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ *  @param moduleID The ID of the module this linker output represents.
+ *
+ *  @param isInternal Whether this is an internal (or an external) module.
+ *
+ *  @param output The [[LinkerOutput]] for this module.
+ *
+ *  @param moduleName Name of this module for import. A [[Linker]] will use
+ *      this name to include this module from one of its dependees. Must be set,
+ *      if this module is internal.
```
Comment (sjrd) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala]: I do not really understand what `moduleName` does. Is this basically the relative path of this module as seen from the modules that will depend on it? I guess that assumes that all modules are in the same directory?

What is the meaning of `moduleName` for an external module?

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+final class ModuleLinkerOutput private (
+    val moduleID: ModuleOutputSpec.ModuleID,
+    val isInternal: Boolean,
+    val output: LinkerOutput,
+    val moduleName: Option[String]
+) {
+  private def this(moduleID: ModuleOutputSpec.ModuleID, output: LinkerOutput) =
+    this(moduleID, false, output, None)
+
+  def withInternal: ModuleLinkerOutput =
```
Comment (sjrd) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala]: If an internal module must have a module name, perhaps `withInternal` should take `moduleName` as parameter?

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleOutputSpec.scala (hunk 5 above)
Comment (sjrd) [reply linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleOutputSpec.scala]: If dependencies must/can be explicitly specified, what happens if they contradict the reality? For example, if I forget to declare a dependency from A to B, but in practice a class in A references a class in B?

If we have to detect those issues, why can't we *always* detect dependencies, making this explicitly configuration redundant?

Comment (gzm0) [reply linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleOutputSpec.scala]: > How does the caller of withDeps create ModuleIDs? It seems there is no public way to construct one.

`Module#apply` creates one from a string.

> If dependencies must/can be explicitly specified, what happens if they contradict the reality?

Does this answer your question: https://github.com/scala-js/scala-js/issues/2681#issuecomment-581153171?

tl;dr IMHO it is important to make sure there are no *unwanted* dependencies. Maybe we should call this `withAllowedStaticDependencies` or something like that.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Linker.scala (hunk 2 above)
Comment (gzm0) [reply linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Linker.scala]: Right... There is actually the additional detail that a module does not need an internal module if no other modules depend on it. This constraint seems to not be easily mapped into the type system. So we need to validate outputs anyways.

That being said, the more type safety the better is of course also an argument :S

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+package org.scalajs.linker.interface
+
+/** A single Module output specification for a linker run.
+ *
+ *  The fields [[moduleID]] and [[isInternal]] uniquely identify this output in
+ *  a [[Linker]] run.
+ *
+ *  @param moduleID The ID of the module this linker output represents.
+ *
+ *  @param isInternal Whether this is an internal (or an external) module.
```
[the remark this answers is not in the document]
Comment (gzm0) [reply linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala]: Ah. An internal module is a module that is to be imported by other SJS emitted modules. An external module is a module that contains definitions exported from SJS.

external modules are always leafs from SJS's POV.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala (hunk 7 above)
Comment (gzm0) [reply linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala]: `moduleName` is the `"module-name"` of the import statement ([MDN][1]). It is used in the modules depending on this one to import it. This is why it is required for internal modules but not external modules. For external modules, this would be ignored but not harmful (thought is that it might simplify upstream code if we allow setting it all the time).

This is similar to the `URI`s in `LinkerOutput`: We allow to set them partially even if it is "typically meaningless".

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala (hunk 8 above)
Comment (gzm0) [reply linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala]: Probably a better design choice indeed. Although see my previous comment about `moduleName` being meaningful all the time and this maybe being unnecessary burden on upstream code.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Linker.scala (hunk 2 above)
Comment (sjrd) [reply linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Linker.scala]: It may not need a module-name, but it still needs an internal module, just so you know what classes to put in what file.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (sjrd) [reply linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala]: Is it then the case that a module is external iff one of the classes it selects contains a top-level export?

What if I have a module that contains a top-level export *and* declares a Scala class that is used by another module? Should that be ruled out?

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala (hunk 7 above)
Comment (sjrd) [reply linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala]: OK, I see.

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala (hunk 8 above)
Comment (sjrd) [reply linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleLinkerOutput.scala]: I'm not sure it helps unless `withInternal` takes a boolean. Because it's a choice between (currently):
```scala
val moduleOutput0 = ModuleLinkerOutput(???)
  .withModuleName(someModuleName)
val moduleOutput =
  if (isInternal) moduleOutput0.withInternal
  else moduleOutput0
```
and (with `withModule` taking the `moduleName`):
```scala
val moduleOutput0 = ModuleLinkerOutput(???)
val moduleOutput =
  if (isInternal) moduleOutput0.withInternal(someModuleName)
  else moduleOutput0
```
It would be simplified if `withInternal` took a boolean instead, as that would allow:
```scala
val moduleOutput = ModuleLinkerOutput(???)
  .withModuleName(someModuleName)
  .withInternal(isInternal)
```
